### ISSUE TITLE
memory dump incorrect on mach-o file

### ISSUE THREAD

**Issue by dbrumley** [BinaryAnalysisPlatform/bap#230]
There appears to be a bug in memory dumping macho files (or the plugin is wrong).

I am using the following plugin to print memory:
```ocaml
open Core_kernel.Std
open Bap.Std
open Format

let print_sections p =
  Project.memory p |> Memmap.to_sequence |> Seq.iter ~f:(fun (mem,x) ->
      Option.iter (Value.get Image.section x) ~f:(fun name ->
          printf "Section: %s@.%a@." name Memory.pp mem))

let () = Project.register_pass' "print-sections" print_sections
```

I compile the following C as a macho file:
```
#include <stdio.h>
#include <ctype.h>

int count(char *str)
{
  int lettercount[26];
  int i, count, l;

  for(i=0; i < 26; i++) lettercount[i] = 0;
  i = 0;
  count = 0;

  while(str[i] != 0){
   l = tolower(str[i]);
   count++;
   if(l >= (int)'a' && l <= (int)'z'){
      lettercount[l-(int)'a'] ++;
   }
   i++;
  }
  for(i =0; i < 26; i++)
    printf("%c: %d ", i+'a', lettercount[i]);
  printf("\n");
  return count;
}

int main(int argc, char *argv[])
{
  if(argc > 1) {
    return count(argv[1]);
  }  else {
    printf("Usage: %s <string>\n", argv[0]);
    printf("\tPrints a count for each letter in <string>\n");
    printf("\tReturns total number of characters counted.\n");
  }
  return 0;
}
```

Memory dump in IDA:
![screen shot 2015-06-05 at 9 19 17 am](https://cloud.githubusercontent.com/assets/2975959/8006820/15e72086-0b65-11e5-91d4-41b8cf4fd2ff.png)

The memory dump in BAP:

```
bash-3.2$ bap -lsimplememory ./exe
Section: __text
100000CF0  CF FA ED FE 07 00 00 01 03 00 00 80 02 00 00 00 |................|
100000D00  10 00 00 00 60 05 00 00 85 00 20 00 00 00 00 00 |....`..... .....|
100000D10  19 00 00 00 48 00 00 00 5F 5F 50 41 47 45 5A 45 |....H...__PAGEZE|
100000D20  52 4F 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |RO..............|
100000D30  00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 |................|
100000D40  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |................|
100000D50  00 00 00 00 00 00 00 00 19 00 00 00 28 02 00 00 |............(...|
100000D60  5F 5F 54 45 58 54 00 00 00 00 00 00 00 00 00 00 |__TEXT..........|
100000D70  00 00 00 00 01 00 00 00 00 10 00 00 00 00 00 00 |................|
100000D80  00 00 00 00 00 00 00 00 00 10 00 00 00 00 00 00 |................|
100000D90  07 00 00 00 05 00 00 00 06 00 00 00 00 00 00 00 |................|
100000DA0  5F 5F 74 65 78 74 00 00 00 00 00 00 00 00 00 00 |__text..........|
100000DB0  5F 5F 54 45 58 54 00 00 00 00 00 00 00 00 00 00 |__TEXT..........|
100000DC0  F0 0C 00 00 01 00 00 00 F2 01 00 00 00 00 00 00 |................|
100000DD0  F0 0C 00 00 04 00 00 00 00 00 00 00 00 00 00 00 |................|
100000DE0  00 04 00 80 00 00 00 00 00 00 00 00 00 00 00 00 |................|
100000DF0  5F 5F 73 74 75 62 73 00 00 00 00 00 00 00 00 00 |__stubs.........|
100000E00  5F 5F 54 45 58 54 00 00 00 00 00 00 00 00 00 00 |__TEXT..........|
100000E10  E2 0E 00 00 01 00 00 00 12 00 00 00 00 00 00 00 |................|
100000E20  E2 0E 00 00 01 00 00 00 00 00 00 00 00 00 00 00 |................|
100000E30  08 04 00 80 00 00 00 00 06 00 00 00 00 00 00 00 |................|
100000E40  5F 5F 73 74 75 62 5F 68 65 6C 70 65 72 00 00 00 |__stub_helper...|
100000E50  5F 5F 54 45 58 54 00 00 00 00 00 00 00 00 00 00 |__TEXT..........|
100000E60  F4 0E 00 00 01 00 00 00 2E 00 00 00 00 00 00 00 |................|
100000E70  F4 0E 00 00 02 00 00 00 00 00 00 00 00 00 00 00 |................|
100000E80  00 04 00 80 00 00 00 00 00 00 00 00 00 00 00 00 |................|
100000E90  5F 5F 63 73 74 72 69 6E 67 00 00 00 00 00 00 00 |__cstring.......|
100000EA0  5F 5F 54 45 58 54 00 00 00 00 00 00 00 00 00 00 |__TEXT..........|
100000EB0  22 0F 00 00 01 00 00 00 79 00 00 00 00 00 00 00 |".......y.......|
100000EC0  22 0F 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |"...............|
100000ED0  02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |................|
100000EE0  5F 5F                                           |__              |

Section: __stubs
100000EE2  CF FA ED FE 07 00 00 01 03 00 00 80 02 00 00 00 |................|
100000EF2  10 00                                           |..              |

Section: __stub_helper
100000EF4  CF FA ED FE 07 00 00 01 03 00 00 80 02 00 00 00 |................|
100000F04  10 00 00 00 60 05 00 00 85 00 20 00 00 00 00 00 |....`..... .....|
100000F14  19 00 00 00 48 00 00 00 5F 5F 50 41 47 45       |....H...__PAGE  |

Section: __cstring
100000F22  CF FA ED FE 07 00 00 01 03 00 00 80 02 00 00 00 |................|
100000F32  10 00 00 00 60 05 00 00 85 00 20 00 00 00 00 00 |....`..... .....|
100000F42  19 00 00 00 48 00 00 00 5F 5F 50 41 47 45 5A 45 |....H...__PAGEZE|
100000F52  52 4F 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |RO..............|
100000F62  00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 |................|
100000F72  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |................|
100000F82  00 00 00 00 00 00 00 00 19 00 00 00 28 02 00 00 |............(...|
100000F92  5F 5F 54 45 58 54 00 00 00                      |__TEXT...       |
```

**Comment by dbrumley** [comment]
I am using bap master from yesterday morning. Forgot to mention that.

**Comment by ivg** [comment]
yes, @rvantonder already found this bug yesterday, so it was fixed yesterday evening and released with 0.9.8. But thanks, for the report. We also need to use our issue tracker more extensively, instead of side channels of communication.
